server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,235 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public class IncrementalBulkService {
+
+    private final boolean allowExplicitIndex;
+    private final Client client;
+
+    public IncrementalBulkService(Settings settings, Client client) {
+        this.allowExplicitIndex = BaseRestHandler.MULTI_ALLOW_EXPLICIT_INDEX.get(settings);
+        this.client = client;
+    }
+
+    public BaseRestHandler.RequestBodyChunkConsumer newBulkRequest(RestRequest request, ActionListener<BulkResponse> listener) {
+        return new BulkChunkHandler(request, listener);
+    }
+
+    private class BulkChunkHandler implements BaseRestHandler.RequestBodyChunkConsumer {
+
+        private final RestRequest request;
+        private final ActionListener<BulkResponse> listener;
+
+        private final Map<String, String> stringDeduplicator = new HashMap<>();
+        private final String defaultIndex;
+        private final String defaultRouting;
+        private final FetchSourceContext defaultFetchSourceContext;
+        private final String defaultPipeline;
+        private final boolean defaultListExecutedPipelines;
+        private final String waitForActiveShards;
+        private final Boolean defaultRequireAlias;
+        private final boolean defaultRequireDataStream;
+        private final TimeValue timeout;
+        private final String refresh;
+        private final BulkRequestParser parser;
+
+        private volatile RestChannel restChannel;
+        private volatile boolean isDone = false;
+        private BulkRequest bulkRequest = null;
+        private int offset = 0;
+        private final ArrayList<ReleasableBytesReference> parsedChunks = new ArrayList<>(4);
+        private final ArrayDeque<ReleasableBytesReference> unParsedChunks = new ArrayDeque<>(4);
+        private final ArrayList<BulkResponse> responses = new ArrayList<>(2);
+
+        private BulkChunkHandler(RestRequest request, ActionListener<BulkResponse> listener) {
+            this.request = request;
+            this.defaultIndex = request.param("index");
+            this.defaultRouting = request.param("routing");
+            this.defaultFetchSourceContext = FetchSourceContext.parseFromRestRequest(request);
+            this.defaultPipeline = request.param("pipeline");
+            this.defaultListExecutedPipelines = request.paramAsBoolean("list_executed_pipelines", false);
+            this.waitForActiveShards = request.param("wait_for_active_shards");
+            this.defaultRequireAlias = request.paramAsBoolean(DocWriteRequest.REQUIRE_ALIAS, false);
+            this.defaultRequireDataStream = request.paramAsBoolean(DocWriteRequest.REQUIRE_DATA_STREAM, false);
+            this.timeout = request.paramAsTime("timeout", BulkShardRequest.DEFAULT_TIMEOUT);
+            this.refresh = request.param("refresh");
+            this.parser = new BulkRequestParser(true, request.getRestApiVersion());
+            this.listener = listener;
```

Comment:
I wonder if we need the listener up front or if we should instead pass it into a final `done(listener)` method, invoked when the rest layer knows we are done.